### PR TITLE
common/blkdev: a few fixes to device name mapping

### PR DESCRIPTION
**Pull request by liewegas**
- map to raw devices when LVM is consuming a parititon
- use a new get_raw_devices() that does everything
- report get_device_id errors
- fix mon to use the new helpers

Fixes http://tracker.ceph.com/issues/37542

### CODE REVIEW

**Comment by wjwithagen** [comment]
@liewegas 
I'm missing the (empy) FreeBSD version for `get_raw_device()`

**Comment by tchaikov** [comment]
```
2019-01-01T03:15:32.537 INFO:teuthology.orchestra.run.smithi050.stderr:2019-01-01 03:15:32.536 7f7e4b6d7700  1 -- 172.21.15.50:41670/3999261857 --> 172.21.15.100:6813/11654 -- comma
nd(tid 0: {"prefix": "mgr self-test module", "target": ["mgr", ""], "module": "devicehealth"}) v1 -- 0x7f7e200014d0 con 0
2019-01-01T03:15:34.403 INFO:tasks.ceph.mon.b.smithi100.stdout: wholedisk of nvme0n1 is nvme0n1
2019-01-01T03:15:34.403 INFO:tasks.ceph.mon.b.smithi100.stderr:2019-01-01 03:15:34.406 7f5e22b1e700 -1 mon.b@1(peon) e1 get_device_id failed on nvme0n1:
2019-01-01T03:15:34.409 INFO:tasks.ceph.mgr.y.smithi100.stderr:2019-01-01 03:15:34.410 7f2d2755d700 -1 Remote method threw exception: Traceback (most recent call last):
2019-01-01T03:15:34.410 INFO:tasks.ceph.mgr.y.smithi100.stderr:  File "/usr/lib/ceph/mgr/devicehealth/module.py", line 189, in self_test
2019-01-01T03:15:34.410 INFO:tasks.ceph.mgr.y.smithi100.stderr:    assert before != after
2019-01-01T03:15:34.410 INFO:tasks.ceph.mgr.y.smithi100.stderr:AssertionError
```

```
2019-01-01 03:15:34.402 7f5e22b1e700  0 mon.b@1(peon) e1 handle_command mon_command({"prefix": "smart", "devid": "INTEL_SSDPEDMD800G4_CVFT65010041800CGN", "format": "json"} v 0) v1
...
2019-01-01 03:15:34.406 7f5e22b1e700 -1 mon.b@1(peon) e1 get_device_id failed on nvme0n1:
```
where `before` was `INTEL_SSDPEDMD800G4_CVFT65010041800CGN`, while `after` was `nvme0n1`.

File: src/common/blkdev.h [hunk 1]
```diff
@@ -29,6 +29,11 @@ extern int block_device_run_smartctl(const char *device, int timeout,
 extern int block_device_get_metrics(const char *device, int timeout,
 				    json_spirit::mValue *result);
 
+// do everything to translate a device to the raw phsyical devices that
```
Comment (neha-ojha) [inline src/common/blkdev.h]: nit: s/phsyical/physical/

File: src/common/blkdev.cc [hunk 2]
```diff
@@ -479,9 +522,122 @@ std::string get_device_id(const std::string& devname)
...
+  int ret = nvmecli.spawn();
+  if (ret != 0) {
+    *result = std::string("error spawning nvme command: ") + nvmecli.err();
+    return ret;
+  }
+
+  bufferlist output;
+  ret = output.read_fd(nvmecli.get_stdout(), 100*1024);
+  if (ret < 0) {
+    // *result = std::string("failed read nvme output: ") + cpp_strerror(-ret);
```
Comment (neha-ojha) [inline src/common/blkdev.cc]: guess we can get rid of this, as we are populating result later

File: src/common/blkdev.h [hunk 3]
```diff
@@ -17,10 +21,16 @@ enum blkdev_prop_t {
 extern int get_device_by_path(const char *path, char* partition, char* device, size_t max);
 
 
-extern std::string get_device_id(const std::string& devname);
+extern std::string get_device_id(const std::string& devname,
+				 std::string *err=0);
 extern void get_dm_parents(const std::string& dev, std::set<std::string> *ls);
-extern int block_device_run_smartctl(const char *device, int timeout,
-				     std::string *result);
+extern int block_device_get_metrics(const string& devname, int timeout,
```
Comment (hsiang41) [inline src/common/blkdev.h]: @liewegas Compile failed. This line should change "string&" to "std::string&"
extern int block_device_get_metrics(const std::string& devname, int timeout,

Comment (neha-ojha) [reply src/common/blkdev.h]: @hsiang41 See https://github.com/ceph/ceph/pull/25783

**Pull request merged**